INFINITY-1677 Implement folder support for cassandra/elastic/hdfs

nickbp opened the pull request:
Updates cassandra/elastic/hdfs to support foldered service names (i.e. names with slashes). Each service needed to be made folder-safe in e.g. hostname references. Adds integration tests for each service to validate that they continue to be folder-safe.

- Adds a new default `FRAMEWORK_HOST` envvar to tasks which can be used for intra-service hostnames, while also ensuring that those hostnames are slash-safe.
- Updates existing hostname references across all services to use `FRAMEWORK_HOST`.
- Updates zk references to be folder-safe as well.
- Adds mode to `TemplateUtils` to throw when template values are missing, rather than silently making them empty. Mainly for HDFS' endpoint template rendering (avoid recurrences of HDFS-461), but may also be useful elsewhere.
- Updates docs to reflect the new `FRAMEWORK_HOST` envvar and other docs cleanup.

ArtRand commented on frameworks/cassandra/tests/config.py:
```diff
@@ -2,32 +2,132 @@
+FOLDERED_SERVICE_VIP_HOST = FOLDERED_SERVICE_NAME.replace('/', '') + '.l4lb.thisdcos.directory'
 DEFAULT_TASK_COUNT = 3
 
+DEFAULT_NODE_ADDRESS = os.getenv('CASSANDRA_NODE_ADDRESS', 'node-0-server.cassandra.autoip.dcos.thisdcos.directory')
```
just an idea, but why not write to a random node?

ArtRand commented on frameworks/elastic/tests/config.py:
```diff
@@ -10,12 +10,16 @@
 import sdk_utils
 
 PACKAGE_NAME = 'elastic'
+FOLDERED_SERVICE_NAME = '/test/integration/' + PACKAGE_NAME
```
this is in two (at least?) tests now, factor it out into `sdk_folders.py` or the like?

ArtRand commented on frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/Main.java:
```diff
@@ -43,25 +49,23 @@ public static void main(String[] args) throws Exception {
     }
 
-    private static String renderTemplate(String pathStr) {
+    private static String renderTemplate(String filename, String serviceName) {
```
👍

ArtRand commented on frameworks/hdfs/tests/test_shakedown.py:
```diff
@@ -24,36 +26,68 @@
 def test_integrity_on_data_node_failure():
     shakedown.wait_for(
-        lambda: write_data_to_hdfs("data-0-node.hdfs.autoip.dcos.thisdcos.directory", TEST_FILE_1_NAME),
+        lambda: write_data_to_hdfs('data-0-node.{}'.format(FOLDERED_SERVICE_AUTOIP_HOST), TEST_FILE_1_NAME),
```
Looking into this, because I'm doing a lot of the same things in the overlay tests. this is an example of when we want to use the `.mesos` DNS because we want the host IP to run the command on. Otherwise, when you call `shakedown.run_command_on_agent(<container_ip>...)` it won't work.

nickbp commented on frameworks/cassandra/tests/test_sanity.py:
```diff
@@ -1,35 +1,44 @@
 
 def teardown_module(module):
-    install.uninstall(PACKAGE_NAME)
+    #install.uninstall(FOLDERED_SERVICE_NAME, package_name=PACKAGE_NAME)
```
It isn't, I left it that way by accident. Uncommented.

nickbp commented on frameworks/cassandra/tests/test_sanity.py:
```diff
@@ -62,20 +79,20 @@ def test_repair_plan_completes():
     key_id = os.getenv('AWS_ACCESS_KEY_ID')
     if not key_id:
-        assert False, 'AWS credentials are required for this test. Disable test with e.g. TEST_TYPES="sanity and not aws"'
+        assert False, 'AWS credentials are required for this test. Disable test with e.g. TEST_TYPES="speedy and not aws"'
```
That was just me accidentally leaving some local changes in the PR. Removed..

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/api/EndpointUtils.java:
```diff
@@ -58,15 +58,26 @@ public static String toEndpoint(String hostname, int port) {
-                removeSlashes(replaceDotsWithDashes(serviceName)),
-                AUTOIP_HOST_TLD);
-        return toEndpoint(hostname, port);
+        return String.format("%s.%s", removeSlashes(replaceDotsWithDashes(serviceName)), AUTOIP_HOST_TLD);
```
There isn't, removed the `EndpointUtils` versions in favor of the `Constants` versions

nickbp commented:
Added a `sdk_hosts` utility for autoip/vip/mesosdns hostnames. Switched tests to use the utility, and switched host system lookups to use the mesosdns host instead of the autoip host.

nickbp commented:
OK found the cause: cassandra's svc.yml was still using scheduler envvars for things which were only available as task envvars. Testing locally worked, lets see if CI catches anything else (particularly any typos from the earlier `sdk_hosts` refactoring), but should otherwise be good to go.

benclarkwood commented:
retest this please (Haven't seen this HDFS flake before, and Loren upped some Elastic timeouts)

nickbp commented:
The failures aren't flakes, in cb0d07b I had changed how hosts were being handled, and that broke something for both of them. Fixed elastic locally (so far, running through it now) and diagnosing what I broke with HDFS.
